cmake/kokkos_arch.cmake

```
@@ -273,8 +274,57 @@ if(KOKKOS_ARCH_NATIVE)
   compiler_specific_flags(COMPILER_ID KOKKOS_CXX_HOST_COMPILER_ID DEFAULT ${KOKKOS_NATIVE_FLAGS})
 endif()
 
+#------------------------------- KOKKOS NEON and SVE detection ---------------------------
+function(kokkos_use_neon_if_compiler_allows_it)
+  unset(KOKKOS_COMPILER_HAS_ARM_NEON CACHE)
+  check_source_compiles(
+    CXX
+    "
+    #include <arm_neon.h>
+    int main() {
+        float32x2_t a;
+        a = vadd_f32(a, a);
+    }
+    "
+    KOKKOS_COMPILER_HAS_ARM_NEON
+  )
+
+  #FIXME_Kokkos_launch_compiler
+  get_property(kokkos_global_rule_compile GLOBAL PROPERTY RULE_LAUNCH_COMPILE)
+  if("${kokkos_global_rule_compile}" MATCHES "kokkos_launch_compiler")
+    message(WARNING "The use of 'kokkos_launch_compiler' prevents reliable NEON detection. Disabling NEON.")
+    set(KOKKOS_COMPILER_HAS_ARM_NEON OFF)
+  endif()
```

Why can't you use `KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA` like we did before?

---

because the detection is at risk for any compiler that we bend towards something else using the wrapper

---

At some point we must also decide between respecting what the user is asking for and forcefully ignoring requests.

---

> At some point we must also decide between respecting what the user is asking for and forcefully ignoring requests.

I have a similar opinion to this, though I don't have a concrete suggestion. Maybe the failure mode should be different between when the user specified an architecture vs when `NATIVE` was used?

---

I am fine either way. I just thought disabling them like this prevents more false positives like with hopper

---

> I have a similar opinion to this, though I don't have a concrete suggestion. Maybe the failure mode should be different between when the user specified an architecture vs when `NATIVE` was used?

I don't think we should fail because NEON/SVE is unsupported. Since the `Kokkos_ARCH_*` options for cpu enable `-march` options, some users who don't care about SIMD could use these CMake options for optimization purpose only, and preventing them to use these options because of something they don't care about might not be a good idea.

---

> because the detection is at risk for any compiler that we bend towards something else using the wrapper

Since the issue comes from nvcc not supporting NEON and SVE when compiling cuda source code, I think we can only disable when nvcc is detected and `Kokkos_ENABLE_CUDA=ON` (or maybe I'm missing something?)

---

I think failing if NEON/SVE simd intrinsics aren't supported is fine. 

> Maybe the failure mode should be different between when the user specified an architecture vs when NATIVE was used?

I agree (even if that's different from what we are currently doing).